### ISSUE TITLE
Privacy and Information Leakage

### ISSUE THREAD

**Issue by DigiTec** [immersive-web/webxr#77]
Even though multiple script contexts will have a separate instance of the VRDisplay, they will share the same backing device. There needs to be some set of controls in place on each API to determine what should and should not be leaked.

Example #1:

``` html
<page>
    <iframe allowvr src="ad_network">
</page>
```

Page begins to present in VR and thus `isPresenting` is true. The iframe also accesses the list of displays and interrogates `isPresenting` and finds it to be true. But the page is not the one presenting and so the advertisement detects they are hosted in page which supports WebVR and that the user is actively in a VR experience. Compare this to full-screen APIs. The ad cannot detect in any case that the hosting page is in full-screen or not and so there is no information leakage.

We probably need to describe some VRDislay device model that can handle this situation. That would include cross tab as well. I just realized that Chrome doesn't constrain access to the Gamepad across a tab boundary, so if I open 2 tabs, press a button in each one, then they can read each others data even when they are not visible. I don't think we want the same for VRDisplay. For instance, someone shouldn't be able to exitPresent on me from a background tab.

**Comment by kearwood** [comment]
In the current Firefox implementation, only one tab at a time can perform VR Presentation for a particular VRDisplay.  If presentation is already active, the RequestPresent returns a promise that fails.  It is still possible for another tab to terminate VR Presentation, but it is not possible to read frames from another tab.  I wouldn't be opposed to blocking ExitPresent from any tab that did not call RequestPresent.  That may prevent some annoying scripts such as spam ads taking over your display.

**Comment by kearwood** [comment]
I suppose this may have an impact on future AR and VR Shell use cases.  I imagine that the mutual exclusion would only apply to VR content.  AR content could be mixed together with other AR content, but should not have access to frame buffer content from others.  Perhaps this isolation could extend to the iframes embedded within a page.

**Comment by kearwood** [comment]
Another consideration is leakage through the VRPose in cases such as gaze based keyboard entry and button selection.  Perhaps we could expose a method that allows content to express that the current pose is security sensitive, similar to how we can mask password entry.  For the duration of the secure VRPose period, the pose would only be accessible from the content that called RequestPresent.

**Comment by toji** [comment]
I've got a [Chrome CL in progress](https://codereview.chromium.org/2329893002/) that enforces similar presentation restrictions to the Firefox behavior described above. Specifically:

- If any page is currently presenting other pages are disallowed from presenting to that device. (Or ANY device, for that matter. That's overly aggressive and shouldn't be specced, but it's not going to hurt too much short term when VR devices are rare.)
- Only the presenting page can end presentation. Otherwise other pages could just call `exitPresent`/`requestPresent` to force you out of whatever you were doing to look at their content. Presentation automatically ends if the page is closed/navigates away.
- When a page is presenting to a device no other pages can read or reset that devices pose/frame data. The reading restriction is to prevent data leakage, the reset restriction is to prevent malicious pages from spamming `resetPose` in the background, which would effectively make your presenting page unusable.

I've also got another [patch in flight](https://codereview.chromium.org/2331223002/) that adds an `allowvr` attribute to iframes, which does pretty much what you'd expect (very closely modeled on `allowfullscreen`)

- If an iframe doesn't have the `allowvr` attribute (or it's parent doesn't, all the way up to the top level frame) calls to `navigator.getVRDisplays()` will reject the promise.
- Devs can also check `navigator.vrEnabled` to see if VR is allowed in that context or not.
- Technically also prevents VR-related events from firing, which is good and what the spec now says should happen, but it's actually a result of a bug in Chrome's code. ;)

I've mentioned in another email that for interactions with trusted UI (system dialogs, keyboards when entering a URL, etc.) that we may want VR blur and focus events that tell the page "Hey, I'm taking away your ability to present and get the pose for a moment. Might want to pause your video/stop animating/etc." Of course that implies that the WebVR content needs to be hidden or reduced to a thumbnail or something because it can no longer headtrack and leaving it up will be uncomfortable.

Sounds like we're generally on the same page, but we should definitely formalize all this behavior.

The AR case does have the potential to complicate things down the road, but I don't know that it's a good idea to stress about that right now. We've got a long way to go before the idea of multiple websites interacting naturally in an AR environment is feasible. In the meantime a model that enforces a single presenter at a time is probably Good Enough.
